src/Columns/ColumnDynamic.cpp

```diff
@@ -0,0 +1,779 @@
+    auto string_variant_discr = variant_info.variant_name_to_discriminator["String"];
+    variant_col.insertManyIntoVariantFrom(string_variant_discr, *tmp_string_column, 0, length);
+}
+
+
+StringRef ColumnDynamic::serializeValueIntoArena(size_t n, DB::Arena & arena, const char *& begin) const
+{
+    /// We cannot use Variant serialization here as it serializes discriminator + value,
+    /// but Dynamic doesn't have fixed mapping discriminator <-> variant type
+    /// as different Dynamic column can have different Variants.
+    /// Instead, we serialize null bit + variant type name (size + bytes) + value.
+    const auto & variant_col = assert_cast<const ColumnVariant &>(*variant_column);
+    auto discr = variant_col.globalDiscriminatorAt(n);
+    StringRef res;
+    UInt8 null_bit = discr == ColumnVariant::NULL_DISCRIMINATOR;
+    if (null_bit)
+    {
+        char * pos = arena.allocContinue(sizeof(UInt8), begin);
+        memcpy(pos, &null_bit, sizeof(UInt8));
+        res.data = pos;
+        res.size = sizeof(UInt8);
+        return res;
+    }
+
+    const auto & variant_name = variant_info.variant_names[discr];
+    size_t variant_name_size = variant_name.size();
+    char * pos = arena.allocContinue(sizeof(UInt8) + sizeof(size_t) + variant_name.size(), begin);
+    memcpy(pos, &null_bit, sizeof(UInt8));
+    memcpy(pos + sizeof(UInt8), &variant_name_size, sizeof(size_t));
+    memcpy(pos + sizeof(UInt8) + sizeof(size_t), variant_name.data(), variant_name.size());
+    res.data = pos;
+    res.size = sizeof(UInt8) + sizeof(size_t) + variant_name.size();
+
+    auto value_ref = variant_col.getVariantByGlobalDiscriminator(discr).serializeValueIntoArena(variant_col.offsetAt(n), arena, begin);
+    res.data = value_ref.data - res.size;
+    res.size += value_ref.size;
+    return res;
+}
+
+const char * ColumnDynamic::deserializeAndInsertFromArena(const char * pos)
+{
+    auto & variant_col = assert_cast<ColumnVariant &>(*variant_column);
+    UInt8 null_bit = unalignedLoad<UInt8>(pos);
+    pos += sizeof(UInt8);
+    if (null_bit)
+    {
+        insertDefault();
+        return pos;
+    }
+
+    /// Read variant type name.
+    const size_t variant_name_size = unalignedLoad<size_t>(pos);
+    pos += sizeof(variant_name_size);
+    String variant_name;
+    variant_name.resize(variant_name_size);
+    memcpy(variant_name.data(), pos, variant_name_size);
+    pos += variant_name_size;
+    /// If we already have such variant, just deserialize it into corresponding variant column.
+    auto it = variant_info.variant_name_to_discriminator.find(variant_name);
+    if (it != variant_info.variant_name_to_discriminator.end())
+    {
+        auto discr = it->second;
+        return variant_col.deserializeVariantAndInsertFromArena(discr, pos);
+    }
+
+    /// If we don't have such variant, add it.
+    auto variant_type = DataTypeFactory::instance().get(variant_name);
```

Comment:
> This seems a bit scary performance wise, right?

Yes it is, but I couldn't find better solution, because we actually have infinite number of possible data types (because of nested types), so as we don't know all the types in advane (as we do in Variant type), we cannot encode it better than just by type names.

And probably we don't need to care a lot about performace of `serializeValueIntoArena/deserializeAndInsertFromArena` as they are used for aggregation and `Dynamic` column is just not really suitable for aggregation.